File: src/midi/synth/sample_bank.cpp
#include "midi/synth/sample_bank.h"

#include <algorithm>

namespace sonare::midi::synth {

bool SampleBank::add_sample(const float* data, size_t n_frames, const SampleDesc& desc,
                            uint32_t* out_index) {
  if (data == nullptr || n_frames == 0) return false;

  const uint32_t start = static_cast<uint32_t>(pool_.size());
  pool_.insert(pool_.end(), data, data + n_frames);

  Sample s;
  s.region.start = start;
  s.region.end = static_cast<uint32_t>(pool_.size());
  // Loops are clamped into the sample, and a loop mode that survives the clamp
  // as an empty span is dropped — SampleReader reads both taps unchecked.
  const uint32_t loop_start = start + std::min(desc.loop_start, static_cast<uint32_t>(n_frames));
  const uint32_t loop_end = start + std::min(desc.loop_end, static_cast<uint32_t>(n_frames));
  s.region.loop_start = loop_start;
  s.region.loop_end = std::max(loop_start, loop_end);
  s.region.loop_mode =
      (desc.loop_mode == 1 || desc.loop_mode == 3) && s.region.loop_end > s.region.loop_start
          ? desc.loop_mode
          : 0;
  s.root_key = static_cast<uint8_t>(std::min<int>(desc.root_key, 127));
  s.fine_tune_cents = desc.fine_tune_cents;
  s.source_rate = desc.source_rate > 0.0 ? desc.source_rate : 0.0;
  samples_.push_back(s);

  if (out_index != nullptr) *out_index = static_cast<uint32_t>(samples_.size() - 1);
  return true;
}

bool SampleBank::add_zone(uint32_t set, const SampleZoneDesc& zone) {
  if (zone.sample_index >= samples_.size()) return false;
  if (zone.key_lo > zone.key_hi || zone.vel_lo > zone.vel_hi) return false;

  if (set >= sets_.size()) sets_.resize(static_cast<size_t>(set) + 1);

  const Sample& s = samples_[zone.sample_index];
  SampleZone z;
  z.region = s.region;
  z.key_lo = zone.key_lo;
  z.key_hi = zone.key_hi;
  z.vel_lo = zone.vel_lo;
  z.vel_hi = zone.vel_hi;
  z.root_key = s.root_key;
  z.tune_cents = s.fine_tune_cents + zone.tune_cents;
  z.gain = zone.gain;
  z.pan_units = zone.pan_units;
  z.source_rate = s.source_rate;
  sets_[set].push_back(z);
  return true;
}

const SampleZone* SampleBank::find(int32_t set, uint8_t key, uint8_t velocity) const noexcept {
  if (set < 0 || static_cast<size_t>(set) >= sets_.size()) return nullptr;
  for (const SampleZone& z : sets_[static_cast<size_t>(set)]) {
    if (key >= z.key_lo && key <= z.key_hi && velocity >= z.vel_lo && velocity <= z.vel_hi) {
      return &z;
    }
  }
  return nullptr;
}

}  // namespace sonare::midi::synth

File: bindings/wasm/src/sample_bank.ts
import { projectModule, type WasmSampleBank } from './project_internal';
import type { SampleDesc, SampleZoneDesc } from './project_types';

/**
 * Host-supplied PCM for the sample synthesis engine: a bank of float samples
 * plus key/velocity keymaps, bound alongside a {@link SynthPatch} whose
 * `engineMode` is `'sample'`.
 *
 * {@link Project.loadSoundFont} is the other door into sampled playback — it
 * parses a container and brings its own generator model. This one takes
 * waveforms a host already has, described by nothing but a keymap, so playing
 * them does not require authoring an SF2. Decoding is the caller's job, and
 * WASM has no host filesystem: the bank takes mono float frames from memory.
 *
 * A bank is built on the control thread and then read as immutable data. Add
 * every sample and zone BEFORE the bounce that binds it starts: the pool is
 * contiguous and moves as it grows, so a sample added while something sounds
 * invalidates the voices reading it.
 *
 * The embind handle is not garbage-collected, so release it with
 * {@link SampleBank.delete} once every bounce that names it has finished.
 *
 * @example
 * ```typescript
 * const bank = new SampleBank();
 * try {
 *   const index = bank.addSample(pcm, { rootKey: 60, sourceRate: 44100 });
 *   bank.addZone({ sampleIndex: index });  // the whole keyboard
 *   const audio = project.bounceWithSynthInstrument(
 *     { engineMode: 'sample', sampleSet: 0, sampleBank: bank },
 *     { totalFrames: 24000 },
 *   );
 * } finally {
 *   bank.delete();
 * }
 * ```
 */
export class SampleBank {
  private readonly native: WasmSampleBank;
  private released = false;

  /**
   * Identity a binding names this bank by, so no raw pointer crosses into JS.
   * Hosts pass the `SampleBank` itself as `sampleBank`.
   *
   * @internal
   */
  readonly nativeId: number;

  /** Create an empty bank. */
  constructor() {
    this.native = new (projectModule().SampleBank)();
    this.nativeId = this.native.id;
  }

  /**
   * Copy mono float frames into the bank and return the new sample's index,
   * which {@link SampleZoneDesc.sampleIndex} names. The frames are copied, so
   * the array may be reused afterwards.
   *
   * Loop points are clamped inside the sample and a loop mode whose loop
   * survives the clamp empty is dropped, so a malformed loop plays as an
   * unlooped sample rather than as a wrap over nothing. An empty array, and a
   * bank that would exceed 67,108,864 sample points, throw.
   *
   * A NaN or Inf frame, `fineTuneCents` or `sourceRate` throws too, and the
   * bank is left unchanged. Such a value is unattributable once stored: the
   * reader's interpolation spreads one bad frame across the whole sustain, and
   * a bad tuning offset renders the voice silent with no error raised.
   */
  addSample(data: Float32Array, desc: SampleDesc = {}): number {
    return this.native.addSample(data, desc);
  }

  /**
   * Append a key/velocity rectangle to a keymap set, creating any sets below
   * it. A patch names a set; the first zone in it covering a note is the one
   * that sounds.
   *
   * Every bound defaults on its own (see {@link SampleZoneDesc}), so an empty
   * rectangle is the whole keyboard at every velocity and narrowing one axis
   * leaves the other whole. A `sampleIndex` the bank does not have, an inverted
   * key or velocity range, and a `setIndex` at or above 4096 all throw.
   */
  addZone(zone: SampleZoneDesc = {}): void {
    const { setIndex, ...rest } = zone;
    this.native.addZone(setIndex ?? 0, rest);
  }

  /** Samples added so far. */
  sampleCount(): number {
    return this.native.sampleCount();
  }

  /** Keymap sets the bank has (one past the highest index used). */
  setCount(): number {
    return this.native.setCount();
  }

  /** Release the underlying WASM object. Idempotent, as the Node facade is. */
  delete(): void {
    if (this.released) {
      return;
    }
    this.released = true;
    this.native.delete();
  }

  /** Alias for {@link SampleBank.delete}, provided for cross-binding (Node) compatibility. */
  destroy(): void {
    this.delete();
  }
}
